primitives/state-machine/src/overlayed_changes.rs

```diff
@@ -224,326 +181,238 @@ impl OverlayedChanges {
-						.insert(extrinsic);
-				}
-			}
-		}
+		self.top.clear_where(|key, _| key.starts_with(prefix), self.extrinsic_index());
 	}
 
+	/// Removes all key-value pairs which keys share the given prefix.
+	///
+	/// Can be rolled back or committed when called inside a transaction
 	pub(crate) fn clear_child_prefix(
 		&mut self,
 		child_info: &ChildInfo,
 		prefix: &[u8],
 	) {
 		let extrinsic_index = self.extrinsic_index();
 		let storage_key = child_info.storage_key();
-		let map_entry = self.prospective.children_default.entry(storage_key.to_vec())
+		let (changeset, info) = self.children.entry(storage_key.to_vec())
 			.or_insert_with(|| (Default::default(), child_info.to_owned()));
-		let updatable = map_entry.1.try_update(child_info);
+		let updatable = info.try_update(child_info);
 		debug_assert!(updatable);
+		changeset.clear_where(|key, _| key.starts_with(prefix), extrinsic_index);
+	}
 
-		for (key, entry) in map_entry.0.iter_mut() {
-			if key.starts_with(prefix) {
-				entry.value = None;
+	/// Returns the current nesting depth of the transaction stack.
+	///
+	/// A value of zero means that no transaction is open and changes are committed on write.
+	pub fn transaction_depth(&self) -> usize {
+		// The top changeset and all child changesets transact in lockstep and are
+		// therefore always at the same transaction depth.
+		self.top.transaction_depth()
+	}
 
-				if let Some(extrinsic) = extrinsic_index {
-					entry.extrinsics.get_or_insert_with(Default::default)
-						.insert(extrinsic);
-				}
-			}
+	/// Start a new nested transaction.
+	///
+	/// This allows to either commit or roll back all changes that where made while this
+	/// transaction was open. Any transaction must be closed by either `rollback_transaction` or
+	/// `commit_transaction` before this overlay can be converted into storage changes.
+	///
+	/// Changes made without any open transaction are committed immediatly.
+	pub fn start_transaction(&mut self) {
+		self.top.start_transaction();
+		for (_, (changeset, _)) in self.children.iter_mut() {
+			changeset.start_transaction();
 		}
+	}
 
-		if let Some((child_committed, _child_info)) = self.committed.children_default.get(storage_key) {
-			// Then do the same with keys from committed changes.
-			// NOTE that we are making changes in the prospective change set.
-			for key in child_committed.keys() {
-				if key.starts_with(prefix) {
-					let entry = map_entry.0.entry(key.clone()).or_default();
-					entry.value = None;
-
-					if let Some(extrinsic) = extrinsic_index {
-						entry.extrinsics.get_or_insert_with(Default::default)
-							.insert(extrinsic);
-					}
-				}
-			}
-		}
+	/// Rollback the last transaction started by `start_transaction`.
+	///
+	/// Any changes made during that transaction are discarded. Returns an error if
+	/// there is no open transaction that can be rolled back.
+	pub fn rollback_transaction(&mut self) -> Result<(), NoOpenTransaction> {
+		self.top.rollback_transaction()?;
```

Comment:
I thought about that but discarded the idea for these reasons: `OverlayChanges` should be an aggregation of `OverlayChanges` so that we can decide later no longer make transaction depth of top and child changes move in lockstep. Also for consistency reasons: With the same argument we could have the transaction depth itself be managed at the `Overlaychanges` level. It is a conscious decision to not do this.

Reply:
> With the same argument we could have the transaction depth itself be managed at the Overlaychanges level

Indeed, that feel better to me too, but its ok this way too (this makes the code look better from some perspective (except enter runtime)).
It is fine if you find it better this way.